crates/loader/src/bindle/assets.rs

```diff
@@ -50,33 +54,56 @@ impl Copier {
 
         let host = create_dir(&base_dst, component).await?;
         let guest = "/".to_string();
-        self.copy_all(parcels, &host).await?;
+        self.copy_all(parcels, &host, allow_transient_write).await?;
 
         Ok(DirectoryMount { host, guest })
     }
 
-    async fn copy_all(&self, parcels: &[Label], dir: impl AsRef<Path>) -> Result<()> {
-        match stream::iter(parcels.iter().map(|p| self.copy(p, &dir)))
-            .buffer_unordered(crate::MAX_PARALLEL_ASSET_PROCESSING)
-            .filter_map(|r| future::ready(r.err()))
-            .map(|e| log::error!("{:?}", e))
-            .count()
-            .await
+    async fn copy_all(
+        &self,
+        parcels: &[Label],
+        dir: impl AsRef<Path>,
+        allow_transient_write: bool,
+    ) -> Result<()> {
+        match stream::iter(
+            parcels
+                .iter()
+                .map(|p| self.copy(p, &dir, allow_transient_write)),
+        )
+        .buffer_unordered(crate::MAX_PARALLEL_ASSET_PROCESSING)
+        .filter_map(|r| future::ready(r.err()))
+        .map(|e| log::error!("{:?}", e))
+        .count()
+        .await
         {
             0 => Ok(()),
             n => bail!("Error copying assets: {} file(s) not copied", n),
         }
     }
 
-    async fn copy(&self, p: &Label, dir: impl AsRef<Path>) -> Result<()> {
+    async fn copy(
+        &self,
+        p: &Label,
+        dir: impl AsRef<Path>,
+        allow_transient_write: bool,
+    ) -> Result<()> {
         let to = dir.as_ref().join(&p.name);
 
         ensure_under(&dir, &to)?;
 
         if to.exists() {
             match check_existing_file(to.clone(), p).await {
-                Ok(true) => return Ok(()),
-                Ok(false) => (),
+                Ok(true) => {
+                    change_file_permission(&to, allow_transient_write).await?;
+                    return Ok(());
+                }
+                Ok(false) => {
+                    // if destination file is read-only, set it to writable first
```

Comment:
This comment sent me in the wrong direction a bit (I thought there was a bug where you forgot to consult `allow_transient_writes`), partly because `check_existing_file` is already confusing!  I think this branch is "file exists but digest doesn't match so it's like corrupted.  In which case this is "set it writeable because we need to overwrite it" I think?  Sort of a "comment says what the code is doing but needs to say *why* instead" issue if that makes sense!

Reply:
I change it to `file exists but digest doesn't match, set it to writable first for further writing`. Do you think it's more clear? Thank you.